### DISCUSSION TITLE
Backend API for external requests

### DISCUSSION THREAD

**Discussion by SinisterSpatula** [DSJAS/DSJAS#31] [Ideas]
Would love to see an API/backend request system that can interact with the bank to add/modify/delete transactions.

This would come in very handy for integrating the bank with browser plugin's or separate webserver.  Like, if I make a spoof purchase on amazn, and I could somehow have my 'purchase' instantly appear in the bank, that would be downright brilliant.

How hard would it be?  How much work would be needed to code it?  Would it be a "post" to a certain file of dsjas and feeding it with credentials + action + options?

Most important features would be (for me):
* Funds available for given account? DSJAS responds with True/False (or maybe an http header code or something? or maybe just make this respond with the current account balance and let the item requesting it do the math instead)
* Add this transaction to this account in this amount with this description, at the current date/time.  DSJAS responds with success/fail.

Bonus/not necessary but nice (for me):
* Modify transaction details.
* Remove transaction (either most recent, or matching provided details, whatever is easiest to code)
* Possibly have a 'pending transactions' if funds available request is received it could be a pending transaction instead of a final one.  (this is probably going overboard. :)

**Comment by ejv2** [comment]
Currently, the remote module API is just public on the site (provided you have an auth token, which is just the PHPSESSID in your browser). See [here](https://github.com/DSJAS/DSJAS/tree/module-api/public/include/api/modules/remote) for some of the APIs and their behaviours that you have available for modules at the moment. Most of these already have functions in the module API to interact easily.

As for APIs for modifying transactions, that might be quite a bit of work as there would need to be some kind of token system to access the API preventing random people from accessing your API and wrecking your database. As a general rule, anything that causes permanent changes to the site needs verification. I could possibly add a system to the admin panel. Perhaps a new class of extension could be API applications with tokens granted?

If you are aware of the GitHub developer tokens system, it could work something like that. Basically, you request a dev token from the admin panel and keep it a secret. Then, you use it to access some priviledged API.

Thoughts?

**Reply by SinisterSpatula** [reply]
Valid concerns.  Mine is only hosted in my lan so I don't mind security or not.  I just want an easy system that I can say "hey bank please do this."  Because the request will be coming from me and not outside my network.  Perhaps it can be optional, disabled in such use cases.  And for others more secure, based on an admin setting.

**Comment by SinisterSpatula** [comment]
@ethanv2 so, I've been working a bit with node.js, express, passport, etc, and building my own bank, and when I did my API in the route/request handler, I simply call checkAuthenticated() before processing the API request.  Perhaps that's all you would need to do as well?  Just a thought.

**Reply by ejv2** [reply]
That is true, but for this to work in PHP you must have the PHPSESSID token in whatever you use to connect. If you use another program other than your browser, this may be annoying for people to do. That is the main problem. From what you said, it sounds like you want to write separate programs that can automate things on thse site. This will be the purpose of extensions once they are written up properly. I'm at the drawing board for a design for them. It's not enormously difficult, but lots of work if I want to do it properly.

So, yes, that is true, but it might make it a bit more difficult for users who want to use the API, hence why I want to avoid it.

**Comment by ejv2** [comment]
I have placed this on the TODO list and will begin working on it shortly.
